src/methods/methods.jl

```diff
@@ -16,21 +16,27 @@ abstract type RotationMethod{T<:RotationType} end
 
 Calculate the criterion of a given `method` with respect to the factor loading matrix `Λ`.
 """
-function criterion end
+criterion(method::RotationMethod, Λ::AbstractMatrix{<:Real}) = criterion_and_gradient!(nothing, method, Λ)
 
 """
-    criterion_and_gradient(method::RotationMethod, Λ::AbstractMatrix{<:Real})
+    criterion_and_gradient!(∇Q::Union{AbstractMatrix{<:Real}, Nothing},
+                            method::RotationMethod, Λ::AbstractMatrix{<:Real})
 
-Calculate the criterion and gradient of a given `method` with respect to the factor loading
-matrix `Λ`.
+Calculate the quality criterion *Q* and the gradient of a given `method`
+with respect to the factor loading matrix `Λ`.
+The gradient is output into `∇Q`.
+The *∇Q* calculation is skipped if `∇Q = nothing`.
 
-Returns a Tuple with the criterion value as the first element and gradient as the second
-element.
+Returns the *Q* criterion value.
 """
-function criterion_and_gradient(method::RotationMethod, Λ::AbstractMatrix)
-    Q = criterion(method, Λ)
-    ∇Q = gradient(Reverse, Base.Fix1(criterion, method), Λ)
-    return Q, ∇Q
+criterion_and_gradient!
+
+# fallback method that uses AutoDiff
+function criterion_and_gradient!(∇Q, method::RotationMethod, Λ::AbstractMatrix)
+    if !isnothing(∇Q)
+        gradient!(Reverse, ∇Q, Base.Fix1(criterion, method), Λ)
+    end
+    return criterion(method, Λ)
 end
```

Comment:
When do we need to call `criterion_and_gradient!` with `∇Q = nothing`? In the rotate function the gradient is always supplied and a user would simply call `criterion` if gradients are not needed. Am I missing something?

Reply:
In this PR, `criterion_and_gradient!()` becomes the only function that has to be implemented for the rotation method, and `criterion()` is just a wrapper that calls `criterion_and_gradient!(nothing, method, L)`. That way we eliminate code duplication between `criterion()` and `criterion_and_gradient!()`; and since Julia will compile specialized version of `criterion_and_gradient!(nothing, method, L)`, it should effectively eliminate the codepaths related to gradient evaluation.

There's one exception, though: `ComponentLoss` only implements `criterion()`, and in that case the autodiff-based fallback `criterion_and_gradient!()` method would be called -- the one we are discussing now.

Reply:
One thing that I don't like with the current state of PR is that if a specific rotation method does not implement neither `criterion()` nor `criterion_and_gradient!()`, calling `criterion_and_gradient!()` would result in a deadlock, since the fallback implementation would call the `criterion()` wrapper, and the wrapper would call the fallback `criterion_and_gradient!()` and so on.

To resolve this, we can change the fallback `criterion_and_gradient!()` to call the internal `criterion_only()` method.
So if the rotation method (i.e. *ComponentLoss*) does not want to implement the full `criterion_and_gradient!()`, it would have to implement `criterion_only()`.

Reply:
Ok, I've updated the PR so that `criterion()` is always a wrapper for `criterion_and_gradient!(nothing, ...)`.
And those *RotationMethod* subtypes, which opt to only implement criterion calculation and rely on autodiff for gradient, would have to implement `criterion_only()` method.

Reply:
I don't really see why can't keep the design as is. 
There are 3 possibilities of implementation: 

1. Implement `criterion` but not `criterion_and_gradient!`. In this case gradient calculation will be handled by the fallback method of `criterion_and_gradient!` using autodiff.
2. Implement `criterion_and_gradient!` but not `criterion`. Then the rotation just uses the implemented method
3. Implement both `criterion` and `criterion_and_gradient!`. Then rotation will use the implemented method and users have the additional convenience of calling `criterion`.

Reply:
Well, you see what have already happened with varimax: *criterion* and *criterion_and_gradient* ended up having different code.
*criterion_and_gradient!* solves code duplication problem, which would also simplify maintenance in the long run. You would not have to check in each PR that the contributor has properly modified the two versions.
Again, from the user perspective *criterion()* will just work as before (calling *criterion_and_gradient!()* under the hood), and there should be no performance penalty.
*criterion_only* is for those so far rare cases, where gradient calculation is done by autodiff. I added it for 2 reasons: a - to avoid deadlocks and provide proper error message when the rotation method does not implement anything, b - to separate internal RotationMethod calls from the public API, and allow user-facing *criterion* implement common logic in the future (e.g. input parameter checks).

Reply:
Actually, I've realized that `criterion_only()` is not needed after all.
For autodiff-based gradients one has to implement only `criterion_and_gradient!(nothing, ...)`, while `criterion_and_gradient!(grad::Vector, ...)` would be the fallback autodiff-based method that calls `criterion_and_gradient!(nothing, ...)` to calculate the gradient.
And in general, it should always be possible to split `criterion_and_gradient!` into 2 separate definitions for `criterion_and_gradient!(nothing, ...)` and `criterion_and_gradient!(grad::Vector, ...)` for a particular rotation method.
So `criterion_and_gradient!()` would be the only function that is needed to implement for a rotation method, and `criterion()` would be the convenience wrapper.

I can update the PR with that changes. I can also rollback and reenable `criterion()` overloading if that's your strong preference.
I just think it is a bit less convenient in the long run.

Reply:
I've replaced `criterion_only(...)` with `criterion_and_gradient!(nothing, ...)`

Reply:
@p-gw So with the last modifications the user can potentially overload the `criterion()` to calculate the criterion of a specific rotation method, but I'd recommend to keep `criterion()` as a wrapper and overload `criterion_and_gradient!()` instead, that's what is written in the updated docs. I have also further optimizations of criterion and gradient calculations for individual rotation methods, but as this PR is already quite large, I'd rather put them in a separate PR when/if this PR is merged.